libusb/os/linux_android_jni.c

```diff
@@ -0,0 +1,1052 @@
+	JNIEnv *jni_env;
+	jobject config;
+	jstring jversion;
+	const char *sversion;
+	struct usbi_string_descriptor string_desc;
+	struct usbi_device_descriptor desc;
+
+	r = android_jni_env(jni, &jni_env);
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	/* int num_configs = device.getConfigurationCount(); */
+	num_configs =
+		(*jni_env)->CallIntMethod(jni_env,
+			device, jni->UsbDevice_getConfigurationCount);
+
+	*strings_len = 2;
+	*strings = malloc(*strings_len);
+	if (!*strings)
+		return LIBUSB_ERROR_NO_MEM;
+	
+	*descriptors_len = LIBUSB_DT_DEVICE_SIZE;
+	*descriptors = malloc(*descriptors_len);
+	if (!*descriptors)
+		return LIBUSB_ERROR_NO_MEM;
+
+	string_desc = (struct usbi_string_descriptor){
+		.bLength = *strings_len,
+		.bDescriptorType = LIBUSB_DT_STRING,
+	};
+	*(struct usbi_string_descriptor *)*strings = string_desc;
+
+	/* parse binary coded decimal version */
+	/* String jversion = device.getVersion(); */
+	jversion =
+		(*jni_env)->CallObjectMethod(jni_env,
+			device, jni->UsbDevice_getVersion);
+
+	sversion = (*jni_env)->GetStringUTFChars(jni_env, jversion, NULL);
+	sscanf(sversion, "%d.%d", &ones, &hundredths);
+	(*jni_env)->ReleaseStringUTFChars(jni_env, jversion, sversion);
+
+	/* Android usb version bug was fixed in commit 608ec66d62647f60c3988922fead33fd7e07755e in Pie */
+	if (jni->Build__VERSION__SDK_INT < jni->Build__VERSION_CODES__P) {
+		/* undo pre-pie bug */
+		tenths = hundredths / 16;
+		hundredths = hundredths % 16;
+	} else {
+		/* bug has been fixed */
+		tenths = hundredths / 10;
+		hundredths = hundredths % 10;
+	}
+	tens = ones / 10;
+	ones = ones % 10;
+
+	version = hundredths | (tenths << 4) | (ones << 8) | (tens << 12);
+
+	desc = (struct usbi_device_descriptor){
+		.bLength = LIBUSB_DT_DEVICE_SIZE,
+		.bDescriptorType = LIBUSB_DT_DEVICE,
+		.bcdUSB = libusb_cpu_to_le16(version),
+		.bDeviceClass =
+			 (*jni_env)->CallIntMethod(jni_env,
+				device, jni->UsbDevice_getDeviceClass),
+		.bDeviceSubClass =
+			(*jni_env)->CallIntMethod(jni_env,
+				device, jni->UsbDevice_getDeviceSubclass),
+		.bDeviceProtocol =
+			(*jni_env)->CallIntMethod(jni_env,
+				device, jni->UsbDevice_getDeviceProtocol),
+		.bMaxPacketSize0 = 8,
+		.idVendor = libusb_cpu_to_le16(
+			(*jni_env)->CallIntMethod(jni_env,
+				device, jni->UsbDevice_getVendorId)),
+		.idProduct = libusb_cpu_to_le16(
+			(*jni_env)->CallIntMethod(jni_env,
+				device, jni->UsbDevice_getProductId)),
+		.bcdDevice = 0xFFFF,
+		.iManufacturer = android_jni_gen_string(jni_env,
+			strings, strings_len,
+			(*jni_env)->CallObjectMethod(jni_env,
+				device, jni->UsbDevice_getManufacturerName)),
+		.iProduct = android_jni_gen_string(jni_env,
+			strings, strings_len,
+			(*jni_env)->CallObjectMethod(jni_env,
+				device, jni->UsbDevice_getProductName)),
+		.iSerialNumber = android_jni_gen_string(jni_env,
+			strings, strings_len,
+			(*jni_env)->CallObjectMethod(jni_env,
+				device, jni->UsbDevice_getSerialNumber)),
```

Comment:
While this part is elegantly constructed, it doesn't work with all the devices.

In practice, `.iManufacturer`/`.iProduct`/`.iSerialNumber` is used in combination with `libusb_get_string_descriptor_ascii` call which sends the actual control request to the device (isn't it?) and if those string indexes doesn't happen to match the indexes in the deal descriptor (which is the case for some of my devices), the control request will return a wrong string descriptor.

As an idea, I was thinking if it is possible to cache `strings` and eventually use it by `libusb_get_string_descriptor**` functions, but at the end, the devices might define some custom string descriptors, except those three and `iInterface`, which should still be available when a USB control request is done, and not nesessary be mmentioned in the descriptor(s).

So as an alternative: is there an objection against asking a real descriptor from the device using a control request?

Reply:
This is unfortunate as the raw descriptor is not at this time available without user-provided permission for the device.  The purpose of this code is to provide for very basic device enumeration without asking for permission yet.

It's been a long time since I've looked at this code, but I recall the intended behavior was to replace the descriptors with the real ones once a connection is opened.  I think your edge case may already be managed, but if it isn't perhaps a simple solution would be to fill the entries with token values and check for those token values later.

Reply:
@Youw does the code path that starts with this comment address your concern sufficiently?  The precise raw descriptors are retrieved from the device when a handle is produced.  The function you express concern over requires a handle to be passed.
https://github.com/libusb/libusb/pull/874/files#diff-e1bd2bd456629177a570d54c2da4e03ff291bd6e3b7dd7f1c6cb320bd57341b4R561 I'm on mobile so the link may be imperfect ... it's line 561 of this same file.

Reply:
I don't believe it is an edge case. Imagine simple device filtering by Vid, Pid and SerialNumber (common case for devices I work with):
```cpp
bool DeviceMatching(libusb_device* device) {
    libusb_device_descriptor desc;
    int err = libusb_get_device_descriptor(device, &desc);
    // handle err

    if (desc.idVendor != required_vendor_id || desc.idProduct != required_product_id)
        return false;

    libusb_device_handle* dev_handle;
    err = libusb_open(device, &dev_handle);
    // handle err

    unsigned char buffer[1024];
    // I'm getting a string other than Serial Number here
    err = libusb_get_string_descriptor_ascii(dev_handle, desc.iSerialNumber, buffer, 1024);
    // handle err

    bool result = strcmp(buffer, required_serial_number);
    libusb_close(device_handle);
    
    return result;
}
```

Reply:
I see the issue now ... as you mentioned you requested the descriptor before opening the device.  Yes, that may fail =(

Would it be better to provide a dummy value that the get_string_descriptors could recognise and use to return an error?

Additionally the backend could detect if permission is already available, and get the correct descriptors if it is, but I feel like that could result in more errors leaking through developers' tests by only partially covering up the issue.  Making a map of generated to real ids has a similar concern.

Reply:
> could result in more errors leaking through developers' tests by only partially covering up the issue

> Making a map of generated to real ids has a similar concern

I share both conserns.